File: mqtt_retrofit/src/main/java/org/sheedon/mqtt/retrofit/FormBody.java
package org.sheedon.mqtt.retrofit;

import com.google.gson.Gson;

import org.sheedon.mqtt.RequestBody;

import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.List;

/**
 * 参数内容主体
 *
 * @Author: sheedon
 * @Email: sheedonsun@163.com
 * @Date: 2020/2/24 16:33
 */
final class FormBody extends RequestBody {

    private final List<String> encodedNames;
    private final List<String> encodedValues;

    private Gson gson;

    private FormBody(List<String> encodedNames, List<String> encodedValues, Gson gson) {
        this.encodedNames = Utils.immutableList(encodedNames);
        this.encodedValues = Utils.immutableList(encodedValues);
        this.gson = gson;

        if (this.gson == null) {
            this.gson = new Gson();
        }
        integrationData();
    }

    private void integrationData() {
        StringBuilder builder = new StringBuilder();
        builder.append("{");

        int size = encodedNames.size();
        for (int index = 0; index < size; index++) {
            String name = encodedNames.get(index);
            String values = encodedValues.get(index);
            builder.append(convert(name, values));
            builder.append(",");
        }

        if (builder.length() > 1) {
            builder.deleteCharAt(builder.length() - 1);
        }
        builder.append("}");

        updateData(builder.toString());
    }

    private String convert(String name, String values) {
        return "\"" + name + "\":" + gson.toJson(values);
    }


    public static final class Builder {
        private final List<String> names = new ArrayList<>();
        private final List<String> values = new ArrayList<>();
        private final Charset charset;
        private Gson gson;

        Builder() {
            this(null);
        }

        Builder(Charset charset) {
            this.charset = charset;
        }

        Builder add(String name, String value) {
            if (name == null) throw new NullPointerException("name == null");
            if (value == null) throw new NullPointerException("value == null");

            names.add(name);
            values.add(value);

//            names.add(HttpUrl.canonicalize(name, FORM_ENCODE_SET, false, false, true, true, charset));
//            values.add(HttpUrl.canonicalize(value, FORM_ENCODE_SET, false, false, true, true, charset));
            return this;
        }

        Builder bindGson(Gson gson) {
            if (gson == null) throw new NullPointerException("gson == null");
            this.gson = gson;
            return this;
        }

        public FormBody build() {
            return new FormBody(names, values, gson);
        }

        Builder addEncoded(String name, String value) {
            return add(name, value);
        }
    }
}

File: retrofit-converters/gson/src/main/java/org/sheedon/retrofit/gson/FormBody.kt
package org.sheedon.retrofit.gson

import org.sheedon.mqtt.retrofit.FormBodyConverter

/**
 * 参数内容主体
 * 基础参数，格式：{"name":"value","name1":"value1"}
 *
 * @Author: sheedon
 * @Email: sheedonsun@163.com
 * @Date: 2022/5/24 22:02
 */
class FormBody(
    private val encodedNames: List<String> = ArrayList(),
    private val encodedValues: List<String> = ArrayList()
) {


    private fun convertToString(): String {
        val builder = StringBuilder()
        builder.append("{")

        val size = encodedNames.size
        for (index in 0 until size) {
            val name = encodedNames[index]
            val values = encodedValues[index]
            builder.append(convert(name, values))
            builder.append(",")
        }

        if (builder.length > 1) {
            builder.deleteCharAt(builder.length - 1)
        }
        builder.append("}")

        return builder.toString()
    }

    private fun convert(name: String, values: String): String {
        return "\"$name\":$values"
    }

    class Builder : FormBodyConverter {
        private val names: MutableList<String> = java.util.ArrayList()
        private val values: MutableList<String> = java.util.ArrayList()
        override fun add(name: String, value: String) {
            names.add(name)
            values.add(value)
        }

        override fun build(): String {
            return FormBody(names, values).convertToString()
        }

        override fun clone() = Builder()
    }

}